zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java

```diff
@@ -453,40 +456,47 @@ public <T extends ExecutableFlowNode> void takeOutgoingSequenceFlows(
   }
 
   /**
-   * Terminates the process instance if the condition for termination is met, i.e. if a runtime
-   * instruction exists that requires the process instance to be terminated after the given element
-   * is completed or terminated.
+   * Executes runtime instructions for the given flow node element if there are any.
    *
    * @param element the flow node element that is being processed
    * @param context the context of the process instance
-   * @return an Either with the context (left) if the process instance is not terminated by this
-   *     call, or Void (right) if it is. If already terminating, returns the context (left) to allow
-   *     the processors to finish terminating elements.
    */
-  public <T extends ExecutableFlowNode>
-      Either<BpmnElementContext, Void> terminateProcessInstanceIfRuntimeInstructionExists(
-          final T element, BpmnElementContext context) {
+  public <T extends ExecutableFlowNode> void executeRuntimeInstructionsIfNeeded(
+      final T element, final BpmnElementContext context) {
+
+    final var runtimeInstructions =
+        stateBehavior.getRuntimeInstructionsForElementId(
+            context.getProcessInstanceKey(), BufferUtil.bufferAsString(element.getId()));
 
-    if (!stateBehavior.shouldTerminateProcessInstance(
-        context.getProcessInstanceKey(), BufferUtil.bufferAsString(element.getId()))) {
-      return Either.left(context);
+    if (runtimeInstructions.isEmpty()) {
+      return;
     }
+
+    // We only have one runtime instruction type for now (termination), so we can directly execute
+    // its logic (interrupt process instance and request termination).
+    // In the future, we might have branches for different instruction types
```

Comment:
🔧 We could safeguard against it by returning early if the type is not TERMINATE
🔧 We should also highlight that we only execute for the first instruction we encounter by extracting the first one from the list here rather than later.